vpn/dialer.go

```diff
@@ -21,49 +24,67 @@ var (
 
 // A TunDialer contains options for obtaining a network connection tunneled
 // through an OpenVPN endpoint. It uses a userspace gVisor virtual device over
-// the raw VPN tunnel.
+// the raw OpenVPN tunnel.
 type TunDialer struct {
+	// Dialer will be passed to the underlying Client constructor.
 	Dialer          DialerContext
-	raw             *RawDialer
+	client          *Client
 	ns1             string
 	ns2             string
 	skipDeviceSetup bool
+	device          *device
+	tun             *netstack.Net
+	mu              sync.Mutex
+
+	// dependency injection to test client start
+	clientStartFn func(context.Context) error
 }
 
 // NewTunDialer creates a new Dialer with the default nameservers (OpenDNS).
-func NewTunDialer(raw *RawDialer) TunDialer {
-	td := TunDialer{
-		raw: raw,
-		ns1: openDNSPrimary,
-		ns2: openDNSSecondary,
+func NewTunDialer(client *Client) *TunDialer {
+	td := &TunDialer{
+		client: client,
+		ns1:    openDNSPrimary,
+		ns2:    openDNSSecondary,
 	}
 	return td
 }
 
-// NewDialerWithNameservers creates a new Dialer with the passed nameservers.
+// NewTunDialerWithNameservers creates a new TunDialer with the passed nameservers.
 // You probably want to pass the nameservers for your own VPN service here.
-func NewTunDialerWithNameservers(raw *RawDialer, ns1, ns2 string) TunDialer {
-	td := TunDialer{
-		raw: raw,
-		ns1: ns1,
-		ns2: ns2,
+func NewTunDialerWithNameservers(client *Client, ns1, ns2 string) *TunDialer {
+	td := &TunDialer{
+		client: client,
+		ns1:    ns1,
+		ns2:    ns2,
 	}
 	return td
 }
 
-// NewDialerFromOptions creates a new Dialer directly from an Options object.
-func NewTunDialerFromOptions(opt *Options) TunDialer {
-	raw := NewRawDialer(opt)
-	td := TunDialer{
-		raw: raw,
-		ns1: openDNSPrimary,
-		ns2: openDNSSecondary,
+// StartNewTunDialerFromOptions creates a new Dialer directly from an Options
+// object. It also starts the underlying client.
+func StartNewTunDialerFromOptions(opt *Options, dialer DialerContext) (*TunDialer, error) {
+	if dialer == nil {
+		return nil, fmt.Errorf("%w: nil dialer", errBadInput)
+
```

Comment:
```suggestion
```

Reply:
(Do we need to care about the `dialer` being `nil`? Isn't passing a `nil` dialer similar to the programmer screwing up? Do you envision a case where a `nil` dialer is not a programming error?)

Reply:
it is a programming error, but I think I preferred to catch it early rather than showing a confusing error when calling Start (that calls dial) in the same function.